File: src/scpn_z_pinch_core/physics/bennett.py
# SPDX-License-Identifier: AGPL-3.0-or-later
# Commercial license available
# © Concepts 1996–2026 Miroslav Šotek. All rights reserved.
# © Code 2020–2026 Miroslav Šotek. All rights reserved.
# ORCID: 0009-0009-3560-0851
# Contact: www.anulum.li | protoscience@anulum.li
# SCPN Z-Pinch Core — Bennett equilibrium model

"""Bennett equilibrium of a z-pinch column (level-0 device physics).

The model is the classical Bennett pinch (W. H. Bennett, Phys. Rev. 45
(1934) 890) in the form reviewed by M. G. Haines, Plasma Phys. Control.
Fusion 53 (2011) 093001, section 2: the integral pressure balance
``mu0 I^2 / (8 pi) = N k (T_e + T_i)``, the Bennett density profile
``n(r) = n_0 / (1 + r^2 / a^2)^2`` with ``n_0 = N / (pi a^2)``, the
azimuthal field of that profile ``B_theta(r) = mu0 I r / (2 pi (a^2 + r^2))``,
and the Alfven speed and transit time of the column evaluated with the
on-axis density. Every quantity is a closed-form evaluation of the
published model on a validated device configuration; nothing here solves
an equilibrium equation, and no value describes a real machine.
"""

from __future__ import annotations

import math
from dataclasses import dataclass
from typing import Final

from scpn_z_pinch_core.parameters import (
    ELEMENTARY_CHARGE_C,
    MU0,
    Discharge,
    PinchColumn,
    require_non_negative,
    require_positive,
)

PROTON_MASS_KG: Final = 1.67262192369e-27
DEUTERON_MASS_KG: Final = 3.3435837724e-27


@dataclass(frozen=True, slots=True)
class BennettEquilibrium:
    """Closed-form Bennett equilibrium of one validated configuration.

    Parameters
    ----------
    scale_radius_m
        Bennett scale radius ``a`` in metres; the validated column radius.
    current_a
        Pinch current ``I`` in amperes.
    line_density_per_m
        Ion line density ``N`` in inverse metres.
    ion_mass_kg
        Ion mass in kilograms used for the mass density.
    temperature_ev
        Bennett temperature ``T`` (equal ion and electron temperatures)
        in electronvolts.
    axis_density_per_m3
        On-axis density ``n_0 = N / (pi a^2)`` in inverse cubic metres.
    axis_pressure_pa
        On-axis kinetic pressure ``p_0 = 2 n_0 k T`` in pascals.
    edge_field_t
        Azimuthal field at the scale radius ``B_theta(a) = mu0 I / (4 pi a)``
        in tesla.
    alfven_speed_m_s
        ``v_A = B_theta(a) / sqrt(mu0 n_0 m_i)`` in metres per second.
    alfven_transit_time_s
        ``tau_A = a / v_A`` in seconds.
    """

    scale_radius_m: float
    current_a: float
    line_density_per_m: float
    ion_mass_kg: float
    temperature_ev: float
    axis_density_per_m3: float
    axis_pressure_pa: float
    edge_field_t: float
    alfven_speed_m_s: float
    alfven_transit_time_s: float

    def density_at(self, radius_m: float) -> float:
        """Evaluate the Bennett density profile.

        Parameters
        ----------
        radius_m
            Radius in metres; finite and non-negative.

        Returns
        -------
        float
            ``n(r) = n_0 / (1 + (r / a)^2)^2`` in inverse cubic metres.

        Raises
        ------
        DeviceConfigurationError
            If the radius is non-finite or negative.
        """
        ratio = _radius_ratio(radius_m, self.scale_radius_m)
        return self.axis_density_per_m3 / ((1.0 + ratio * ratio) ** 2)

    def pressure_at(self, radius_m: float) -> float:
        """Evaluate the kinetic pressure profile.

        Parameters
        ----------
        radius_m
            Radius in metres; finite and non-negative.

        Returns
        -------
        float
            ``p(r) = p_0 / (1 + (r / a)^2)^2`` in pascals.

        Raises
        ------
        DeviceConfigurationError
            If the radius is non-finite or negative.
        """
        ratio = _radius_ratio(radius_m, self.scale_radius_m)
        return self.axis_pressure_pa / ((1.0 + ratio * ratio) ** 2)

    def field_at(self, radius_m: float) -> float:
        """Evaluate the azimuthal field of the Bennett profile.

        Parameters
        ----------
        radius_m
            Radius in metres; finite and non-negative.

        Returns
        -------
        float
            ``B_theta(r) = mu0 I r / (2 pi (a^2 + r^2))`` in tesla.

        Raises
        ------
        DeviceConfigurationError
            If the radius is non-finite or negative.
        """
        _radius_ratio(radius_m, self.scale_radius_m)
        return (
            MU0
            * self.current_a
            * radius_m
            / (2.0 * math.pi * (self.scale_radius_m**2 + radius_m**2))
        )

    def enclosed_current_at(self, radius_m: float) -> float:
        """Evaluate the current enclosed inside a radius.

        Parameters
        ----------
        radius_m
            Radius in metres; finite and non-negative.

        Returns
        -------
        float
            ``I(r) = I r^2 / (a^2 + r^2)`` in amperes.

        Raises
        ------
        DeviceConfigurationError
            If the radius is non-finite or negative.
        """
        _radius_ratio(radius_m, self.scale_radius_m)
        return self.current_a * radius_m**2 / (self.scale_radius_m**2 + radius_m**2)

    def to_record(self) -> dict[str, float]:
        """Project the equilibrium to a JSON-serialisable record.

        Returns
        -------
        dict[str, float]
            Every field of the equilibrium under its SI-suffixed name.
        """
        return {
            "scale_radius_m": self.scale_radius_m,
            "current_a": self.current_a,
            "line_density_per_m": self.line_density_per_m,
            "ion_mass_kg": self.ion_mass_kg,
            "temperature_ev": self.temperature_ev,
            "axis_density_per_m3": self.axis_density_per_m3,
            "axis_pressure_pa": self.axis_pressure_pa,
            "edge_field_t": self.edge_field_t,
            "alfven_speed_m_s": self.alfven_speed_m_s,
            "alfven_transit_time_s": self.alfven_transit_time_s,
        }


def _radius_ratio(radius_m: float, scale_radius_m: float) -> float:
    """Validate a profile radius and return ``r / a``.

    Parameters
    ----------
    radius_m
        Radius in metres.
    scale_radius_m
        Bennett scale radius in metres.

    Returns
    -------
    float
        The dimensionless radius.

    Raises
    ------
    DeviceConfigurationError
        If the radius is non-finite or negative.
    """
    require_non_negative("radius_m", radius_m)
    return radius_m / scale_radius_m


def bennett_equilibrium(
    column: PinchColumn, discharge: Discharge, ion_mass_kg: float
) -> BennettEquilibrium:
    """Evaluate the Bennett equilibrium of a validated configuration.

    Parameters
    ----------
    column
        Validated pinch-column geometry; its radius is the Bennett scale
        radius ``a``.
    discharge
        Validated discharge; its current and line density enter the
        pressure balance.
    ion_mass_kg
        Ion mass in kilograms; strictly positive (see
        :data:`PROTON_MASS_KG` and :data:`DEUTERON_MASS_KG`).

    Returns
    -------
    BennettEquilibrium
        The closed-form equilibrium quantities.

    Raises
    ------
    DeviceConfigurationError
        If the ion mass is non-finite or not strictly positive.

    Notes
    -----
    The temperature is the equal-species Bennett temperature already
    exposed by :meth:`Discharge.bennett_temperature_ev`, so the two
    surfaces can never disagree. Operation order is fixed so that the
    native kernels reproduce every value bit for bit.
    """
    require_positive("ion_mass_kg", ion_mass_kg)
    scale_radius_m = column.column_radius_m
    current_a = discharge.peak_current_ma * 1.0e6
    line_density_per_m = discharge.ion_line_density_per_m
    temperature_ev = discharge.bennett_temperature_ev()
    axis_density_per_m3 = line_density_per_m / (math.pi * scale_radius_m**2)
    axis_pressure_pa = 2.0 * axis_density_per_m3 * temperature_ev * ELEMENTARY_CHARGE_C
    edge_field_t = MU0 * current_a / (4.0 * math.pi * scale_radius_m)
    alfven_speed_m_s = edge_field_t / math.sqrt(MU0 * axis_density_per_m3 * ion_mass_kg)
    alfven_transit_time_s = scale_radius_m / alfven_speed_m_s
    return BennettEquilibrium(
        scale_radius_m=scale_radius_m,
        current_a=current_a,
        line_density_per_m=line_density_per_m,
        ion_mass_kg=ion_mass_kg,
        temperature_ev=temperature_ev,
        axis_density_per_m3=axis_density_per_m3,
        axis_pressure_pa=axis_pressure_pa,
        edge_field_t=edge_field_t,
        alfven_speed_m_s=alfven_speed_m_s,
        alfven_transit_time_s=alfven_transit_time_s,
    )

File: rust/src/bennett.rs
// SPDX-License-Identifier: AGPL-3.0-or-later
// Commercial license available
// © Concepts 1996–2026 Miroslav Šotek. All rights reserved.
// © Code 2020–2026 Miroslav Šotek. All rights reserved.
// ORCID: 0009-0009-3560-0851
// Contact: www.anulum.li | protoscience@anulum.li
// SCPN Z-Pinch Core — Bennett equilibrium kernel

//! Bennett equilibrium of a z-pinch column (Bennett 1934; Haines 2011 §2),
//! operation-for-operation identical to
//! `scpn_z_pinch_core.physics.bennett.bennett_equilibrium`.

use crate::{ELEMENTARY_CHARGE_C, MU0};

/// Closed-form Bennett equilibrium quantities (SI units in the names).
#[derive(Debug, Clone, Copy, PartialEq)]
pub struct BennettEquilibrium {
    /// Pinch current `I` in amperes.
    pub current_a: f64,
    /// Equal-species Bennett temperature in electronvolts.
    pub temperature_ev: f64,
    /// On-axis density `n_0 = N / (pi a^2)`.
    pub axis_density_per_m3: f64,
    /// On-axis pressure `p_0 = 2 n_0 k T`.
    pub axis_pressure_pa: f64,
    /// Azimuthal field at the scale radius `mu0 I / (4 pi a)`.
    pub edge_field_t: f64,
    /// Alfven speed `B_theta(a) / sqrt(mu0 n_0 m_i)`.
    pub alfven_speed_m_s: f64,
    /// Alfven transit time `a / v_A`.
    pub alfven_transit_time_s: f64,
}

/// Evaluate the Bennett equilibrium.
///
/// Inputs are the validated configuration values (validation lives in the
/// Python floor; the kernel assumes finite, strictly positive inputs).
#[must_use]
pub fn bennett_equilibrium(
    column_radius_m: f64,
    peak_current_ma: f64,
    ion_line_density_per_m: f64,
    ion_mass_kg: f64,
) -> BennettEquilibrium {
    let current_a = peak_current_ma * 1.0e6;
    let temperature_ev = MU0 * (current_a * current_a)
        / (16.0 * std::f64::consts::PI * ion_line_density_per_m * ELEMENTARY_CHARGE_C);
    let axis_density_per_m3 =
        ion_line_density_per_m / (std::f64::consts::PI * (column_radius_m * column_radius_m));
    let axis_pressure_pa = 2.0 * axis_density_per_m3 * temperature_ev * ELEMENTARY_CHARGE_C;
    let edge_field_t = MU0 * current_a / (4.0 * std::f64::consts::PI * column_radius_m);
    let alfven_speed_m_s = edge_field_t / (MU0 * axis_density_per_m3 * ion_mass_kg).sqrt();
    let alfven_transit_time_s = column_radius_m / alfven_speed_m_s;
    BennettEquilibrium {
        current_a,
        temperature_ev,
        axis_density_per_m3,
        axis_pressure_pa,
        edge_field_t,
        alfven_speed_m_s,
        alfven_transit_time_s,
    }
}

/// Bennett density profile `n_0 / (1 + (r / a)^2)^2`.
#[must_use]
pub fn density_at(axis_density_per_m3: f64, scale_radius_m: f64, radius_m: f64) -> f64 {
    let ratio = radius_m / scale_radius_m;
    let inner = 1.0 + ratio * ratio;
    axis_density_per_m3 / (inner * inner)
}

/// Azimuthal field of the Bennett profile `mu0 I r / (2 pi (a^2 + r^2))`.
#[must_use]
pub fn field_at(current_a: f64, scale_radius_m: f64, radius_m: f64) -> f64 {
    MU0 * current_a * radius_m
        / (2.0 * std::f64::consts::PI * (scale_radius_m * scale_radius_m + radius_m * radius_m))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn pressure_balance_closes() {
        let e = bennett_equilibrium(0.005, 0.1, 1.0e18, 3.343_583_772_4e-27);
        let magnetic = 2.0 * e.edge_field_t * e.edge_field_t / MU0;
        assert!((e.axis_pressure_pa - magnetic).abs() <= 1.0e-15 * magnetic);
        assert_eq!(e.current_a, 1.0e5);
    }

    #[test]
    fn profile_and_field_at_the_scale_radius() {
        let e = bennett_equilibrium(0.005, 0.1, 1.0e18, 3.343_583_772_4e-27);
        let quarter = density_at(e.axis_density_per_m3, 0.005, 0.005);
        assert!((quarter - e.axis_density_per_m3 / 4.0).abs() <= 1.0e-15 * quarter);
        let edge = field_at(e.current_a, 0.005, 0.005);
        assert!((edge - e.edge_field_t).abs() <= 1.0e-15 * edge);
        assert_eq!(field_at(e.current_a, 0.005, 0.0), 0.0);
    }
}
